query-engine/driver-adapters/executor/src/driver-adapters-manager/d1.ts

```diff
@@ -52,3 +54,61 @@ export class D1Manager implements DriverAdaptersManager {
     await this.#dispose()
   }
 }
+
+async function migrateDiff(D1_DATABASE: D1Database, migrationScript: string) {
+  // Note: when running a script with multiple statements, D1 fails with
+  // `D1_ERROR: A prepared SQL statement must contain only one statement.`
+  // We thus need to run each statement separately, splitting the script by `;`.
+  const sqlStatements = migrationScript.split(';')
+  const preparedStatements = sqlStatements.map((sqlStatement) => D1_DATABASE.prepare(sqlStatement))
+  await D1_DATABASE.batch(preparedStatements)
+}
+
+async function migrateReset(D1_DATABASE: D1Database) {
+  let { results: rawTables } = ((await D1_DATABASE.prepare(`PRAGMA main.table_list;`).run()) as D1Result)
+  let tables = S
+    .decodeUnknownSync(D1Tables, { onExcessProperty: 'preserve' })(rawTables)
+    .filter((item) => !['_cf_KV', 'sqlite_schema'].includes(item.name))
+
+  const batch = [] as string[]
+
+  // temporarily allow violating foreign key constraints
+  batch.push(`PRAGMA defer_foreign_keys = ${1};`)
+  batch.push(`PRAGMA foreign_keys = ${0};`)
```

Comment:
Based on this and what I tried a couple of weeks ago. `batch.push(`PRAGMA foreign_keys = ${0};`)` will not do anything here and is not needed.
>Because D1 runs every query inside an implicit transaction, user queries cannot change this during a query or migration.

Reply:
applies to the next use down too